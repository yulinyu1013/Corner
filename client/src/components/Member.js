import React from 'react';
import '../styles/memberlist.css';
import 'antd/dist/antd.css';
import { FaRocketchat } from 'react-icons/fa';
import { useHistory } from 'react-router-dom';
import { Menu, Dropdown, message } from 'antd';
import {
  MoreOutlined, UserOutlined, UserDeleteOutlined, UserAddOutlined, CrownOutlined,
} from '@ant-design/icons';

function Member({
  currentUser, user, deleteUser, setAdmin, revokeAdmin, isLoggedIn, isChat,
}) {
  const handleMenuClick = (e) => {
    console.log(currentUser);
    console.log(user);
    if (currentUser.id === user.id && user.level === 3) {
      message.info('You cannot operate on creator.');
      return;
    }
    if (currentUser.level >= user.level && currentUser.level > 1) {
      if (e.key === '1') {
        message.info('OK! The user has been deleted from the group');
        deleteUser(user.id);
      } else if (e.key === '2') {
        message.info('OK! set the user to be an admin');
        setAdmin(user.id);
      } else {
        message.info('OK! revoke the admin role of users');
        revokeAdmin(user.id);
      }
      console.log('click', e);
    } else {
      message.info('Sorry, you dont have such authorization!');
    }
  };

  const history = useHistory();

  const redirectToPrivateChat = () => {
    console.log(currentUser.name, user.name);
    history.push(`/pChat/${currentUser.name}/${user.name}`);
  };

  const menu = (
    <Menu onClick={handleMenuClick}>
      <Menu.Item key="1" icon={<UserDeleteOutlined />}>
        delete member
      </Menu.Item>
      <Menu.Item key="2" icon={<UserAddOutlined />}>
        set admin
      </Menu.Item>
      <Menu.Item key="3" icon={<UserOutlined />}>
        revoke admin
      </Menu.Item>
    </Menu>
  );

  const titleDisplay = (duser) => {
    if (duser.level === 3) {
      return (
        <div>
          <span>
            {duser.name}
            &nbsp;
          </span>
          <CrownOutlined />
        </div>
      );
    }
    if (duser.level === 2) {
      return (
        <div>
          <span>
            {duser.name}
            &nbsp;
          </span>
          <UserOutlined />
        </div>
      );
    }
    return <span>{duser.name}</span>;
  };

  return (
    <div className="member">
      <div className="member-img-info">
        <div className="member-img-container">
          <img className="member-img" src={user.avatar} alt="img" />
        </div>
        <div className="member-info">
          <div className="member-name">
            {titleDisplay(user)}
          </div>
        </div>
      </div>
      <div>
        { !isChat && isLoggedIn && <FaRocketchat className="chat-icon" onClick={() => redirectToPrivateChat()} />}
      </div>
      <div>
        { user.id !== currentUser.id && !isChat
        && (
          <Dropdown.Button
            overlay={menu}
            icon={<MoreOutlined style={{ fontSize: '15px', color: 'black' }} />}
          />
        )}
      </div>
    </div>
  );
}

export default Member;
