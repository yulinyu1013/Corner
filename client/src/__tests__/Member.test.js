import React from 'react';
import renderer from 'react-test-renderer';
import Member from '../components/Member';

const userLevel1 = {
  id: 0,
  name: 'dummy',
  avatar: null,
  level: 1,
}

const currentUserLevel3 = {
  id: 1,
  name: 'dummy',
  avatar: null,
  level: 3,
}

const currentUserLevel1 = {
  id: 2,
  name: 'dummy',
  avatar: null,
  level: 3,
}

const userLevel2 = {
  id: 3,
  name: 'dummy',
  avatar: null,
  level: 1,
}

const currentUserLevelsame = {
  id: 4,
  name: 'dummy',
  avatar: null,
  level: 3,
}

const userLevelsame = {
  id: 4,
  name: 'dummy',
  avatar: null,
  level: 3,
}

it('renders Member correctly when user is admin', () => {
  const tree = renderer.create(<Member currentUser={currentUserLevel3} user={userLevel1} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders Member correctly when user is normal user', () => {
  const tree = renderer.create(<Member currentUser={currentUserLevel1} user={userLevel2} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders Member correctly when user is the currentuser', () => {
  const tree = renderer.create(<Member currentUser={currentUserLevelsame} user={userLevelsame} />).toJSON();
  expect(tree).toMatchSnapshot();
});
