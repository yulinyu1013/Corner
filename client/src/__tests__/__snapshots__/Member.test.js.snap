// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Member correctly when user is admin 1`] = `
<div
  className="member"
>
  <div
    className="member-img-info"
  >
    <div
      className="member-img-container"
    >
      <img
        alt="img"
        className="member-img"
        src={null}
      />
    </div>
    <div
      className="member-info"
    >
      <div
        className="member-name"
      >
        <span>
          dummy
        </span>
      </div>
    </div>
  </div>
  <div />
  <div>
    <div
      className="ant-btn-group ant-dropdown-button"
    >
      <button
        className="ant-btn ant-btn-default"
        onClick={[Function]}
        type="button"
      />
      <button
        className="ant-btn ant-btn-default ant-btn-icon-only ant-dropdown-trigger"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        type="button"
      >
        <span
          aria-label="more"
          className="anticon anticon-more"
          role="img"
          style={
            Object {
              "color": "black",
              "fontSize": "15px",
            }
          }
        >
          <svg
            aria-hidden="true"
            data-icon="more"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`renders Member correctly when user is normal user 1`] = `
<div
  className="member"
>
  <div
    className="member-img-info"
  >
    <div
      className="member-img-container"
    >
      <img
        alt="img"
        className="member-img"
        src={null}
      />
    </div>
    <div
      className="member-info"
    >
      <div
        className="member-name"
      >
        <span>
          dummy
        </span>
      </div>
    </div>
  </div>
  <div />
  <div>
    <div
      className="ant-btn-group ant-dropdown-button"
    >
      <button
        className="ant-btn ant-btn-default"
        onClick={[Function]}
        type="button"
      />
      <button
        className="ant-btn ant-btn-default ant-btn-icon-only ant-dropdown-trigger"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        type="button"
      >
        <span
          aria-label="more"
          className="anticon anticon-more"
          role="img"
          style={
            Object {
              "color": "black",
              "fontSize": "15px",
            }
          }
        >
          <svg
            aria-hidden="true"
            data-icon="more"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`renders Member correctly when user is the currentuser 1`] = `
<div
  className="member"
>
  <div
    className="member-img-info"
  >
    <div
      className="member-img-container"
    >
      <img
        alt="img"
        className="member-img"
        src={null}
      />
    </div>
    <div
      className="member-info"
    >
      <div
        className="member-name"
      >
        <div>
          <span>
            dummy
            Â 
          </span>
          <span
            aria-label="crown"
            className="anticon anticon-crown"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="crown"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M899.6 276.5L705 396.4 518.4 147.5a8.06 8.06 0 00-12.9 0L319 396.4 124.3 276.5c-5.7-3.5-13.1 1.2-12.2 7.9L188.5 865c1.1 7.9 7.9 14 16 14h615.1c8 0 14.9-6 15.9-14l76.4-580.6c.8-6.7-6.5-11.4-12.3-7.9zm-126 534.1H250.3l-53.8-409.4 139.8 86.1L512 252.9l175.7 234.4 139.8-86.1-53.9 409.4zM512 509c-62.1 0-112.6 50.5-112.6 112.6S449.9 734.2 512 734.2s112.6-50.5 112.6-112.6S574.1 509 512 509zm0 160.9c-26.6 0-48.2-21.6-48.2-48.3 0-26.6 21.6-48.3 48.2-48.3s48.2 21.6 48.2 48.3c0 26.6-21.6 48.3-48.2 48.3z"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div />
  <div />
</div>
`;
